oauth2-proxy/oauth2-proxy
# Azure Provider: return to using token's groups claim

## Description

Return to using the groups claim in the Azure provider. This involved the following changes:

1. Switch to using the [Microsoft Graph SDK for Go](https://github.com/microsoftgraph/msgraph-sdk-go) for graph API queries.
2. Use the `groups` claim if it's present.
3. Only query the Microsoft Graph API in the following cases:
   1. `--azure-graph-group-field` is `displayName`. It's only possible to return group IDs in a groups claim (see https://github.com/MicrosoftDocs/azure-docs/issues/38625), so we need to hit the graph API to get their displayName values. This requires the _GroupMember.Read.All_ delegated API permission to be granted to application.
   2. The number of groups is over 200, i.e., group overage. See [how to handle group overage](https://learn.microsoft.com/en-us/azure/active-directory/external-identities/customers/how-to-web-app-role-based-access-control#handle-groups-overage) and [group overages](https://learn.microsoft.com/en-us/security/zero-trust/develop/configure-tokens-group-claims-app-roles#group-overages).

Using this approach, most situations will only require the *User.Read* Microsoft Graph API delegated permission. If it's desired to return display names instead of group IDs (3.i above), the *GroupMember.Read.All* permission is required.

Also update the Azure provider docs:
- general readability improvements
- remove step to modify app manifest for v2.0 as this is not needed (see https://github.com/oauth2-proxy/oauth2-proxy/issues/1666#issuecomment-1333879118 for confusion here)
- add `--azure-graph-group-field` to documentation (added in #1574 but not documented)
- remove `--azure-tenant` from the documentation as this is not necessary because the OIDC discovery URL has what we need

Fixes #1868 #2154

## Motivation and Context

#1574 added support for Azure OAuth 2.0 and querying of groups from the Microsoft Graph API. 

The new groups query that #1574 added cannot be disabled and requires a new API permission, [**GroupMember.Read.All**](https://learn.microsoft.com/en-us/graph/permissions-reference#delegated-permissions-39). This permission requires [administrator consent](https://learn.microsoft.com/en-us/azure/active-directory/develop/permissions-consent-overview?WT.mc_id=Portal-Microsoft_AAD_RegisteredApps#administrator-consent) that would be unnecessary if the groups came from a claim in the ID token. The requirement for administrator consent is particularly painful when using the Azure provider in a multi-tenant scenario, which requires administrators of each tenant to grant consent before their users can use the application.

https://github.com/oauth2-proxy/oauth2-proxy/issues/1666#issuecomment-1297160824 and #1868  show examples of the confusion created by this new requirement.

Additionally, since the groups claim is unused, users who want to limit the groups in the token to only those assigned to the application cannot do so in the current state of the application because all of the user's groups are always returned (see #2154).

## How Has This Been Tested?





1.  [Create an Azure app registration](https://aad.portal.azure.com/#view/Microsoft_AAD_RegisteredApps/CreateApplicationBlade/quickStartType~/null/isMSAApp~/false) with the following settings (do not modify any other settings or grant new API admin permissions):
    - **Name**: Test App without Admin Consent
    - **Supported account types:** Accounts in any organizational directory (Any Azure AD directory - Multitenant)
    - **Redirect URI**: `http://localhost:4180/oauth2/callback`
2. Set client ID and secret environment variables.

    ```bash
    export OAUTH2_PROXY_CLIENT_ID="<client-id>"
    export OAUTH2_PROXY_CLIENT_SECRET="<client-secret>"
    ```

3. Compile and run the proxy with the following settings:

    ```bash
    ./oauth2-proxy \
       --oidc-issuer-url=https://login.microsoftonline.com/organizations/v2.0 \
       --insecure-oidc-skip-issuer-verification \
       --cookie-secret "$(openssl rand -hex 16)" \
       --email-domain "*" \
       --provider azure \
       --redirect-url http://localhost:4180/oauth2/callback
    ```

4. Ensure login at http://localhost:4180 is successful.
5. Restart the proxy with the `--allowed-group "Testing Groups"` and `--azure-graph-group-field "displayName"` flags.

    ```bash
    ./oauth2-proxy \
       --oidc-issuer-url=https://login.microsoftonline.com/organizations/v2.0 \
       --insecure-oidc-skip-issuer-verification \
       --cookie-secret "$(openssl rand -hex 16)" \
       --email-domain "*" \
       --provider azure \
       --redirect-url http://localhost:4180/oauth2/callback \
       --allowed-group "Testing Groups" \
       --azure-graph-group-field "displayName"
    ```

6. Navigate to http://localhost:4180 to initiate the login flow again. This time, a 403 is shown because group validation has been requested, but there are no groups in the returned token (yet).

    ![image](https://github.com/oauth2-proxy/oauth2-proxy/assets/16367391/396631e1-3aed-4aa1-8ce5-4c21e86e1859)

7. Go to *Token Configuration* in the app and add a groups claim for all groups. See the `OVERVIEW.md` file for specific details.

8. Pick a group ID that your user is a member of and update the command above, removing `--azure-graph-group-field "displayName"`.

    ```bash
    ./oauth2-proxy \
       --oidc-issuer-url=https://login.microsoftonline.com/organizations/v2.0 \
       --insecure-oidc-skip-issuer-verification \
       --cookie-secret "$(openssl rand -hex 16)" \
       --email-domain "*" \
       --provider azure \
       --redirect-url http://localhost:4180/oauth2/callback \
       --allowed-group "39cb3e42-99b7-4404-8747-a515f5865277"
    ```

7. As an administrator, grant the **GroupMember.Read.All** API permission to the app created in step 1 (see the updated doc in this revision for instructions).
10. Navigate to http://localhost:4180 to initiate the login flow again. This time, a 403 error is return, indicating that the user is not in the list of allowed groups.

    ![image](https://user-images.githubusercontent.com/16367391/212447842-203a0f3e-782b-40c9-9b18-c590b4499bb9.png)

11. Change the `--allowed-group` flag to a valid Azure group **ID** that I am a member of and restart the proxy.
12. Ensure login at http://localhost:4180 is successful.

## Checklist:
